tools/content-generator/src/main/java/org/projectnessie/tools/contentgenerator/cli/DetachHistory.java

```diff
@@ -0,0 +1,180 @@
+import org.projectnessie.error.NessieConflictException;
+import org.projectnessie.error.NessieNotFoundException;
+import org.projectnessie.error.NessieReferenceConflictException;
+import org.projectnessie.model.Branch;
+import org.projectnessie.model.Content;
+import org.projectnessie.model.ContentKey;
+import org.projectnessie.model.Reference;
+import org.projectnessie.model.Tag;
+import picocli.CommandLine;
+import picocli.CommandLine.Command;
+
+@Command(
+    name = "detach-history",
+    mixinStandardHelpOptions = true,
+    description =
+        "Get the latest contents from one or all branches, commit it directly on top of the root commit (no history), "
+            + "then reassign the original branch(es) to the new commit.")
+public class DetachHistory extends RefreshContent {
+
+  @CommandLine.Option(
+      names = {"--src-suffix"},
+      description =
+          "Name suffix for tagging the original (source) branches, defaults to '-original'.")
+  private String origSuffix = "-original";
+
+  @CommandLine.Option(
+      names = {"--final-suffix"},
+      description =
+          "Name suffix for tagging the the final stat of source branches (equivalent to the tag generated "
+              + "from the --src-suffix option). This suffix is used during retries, defaults to '-completed'.")
+  private String completedSuffix = "-completed";
+
+  @CommandLine.Option(
+      names = {"--max-attempts"},
+      description =
+          "Max attempts (in case of concurrent commits from other actors), defaults to 100.")
+  private int maxAttempts = 100;
+
+  private final Map<String, Branch> sources = new HashMap<>();
+  private final Map<String, Branch> targets = new HashMap<>();
+  private String tmpSuffix;
+  private String rootHash;
+
+  @Override
+  public void execute() throws BaseNessieClientServerException {
+    // Overview:
+    // 1) Tag old source branch HEADs (in setup()).
+    // 2) Copy the latest content to temporary branches.
+    // 3) Reassign source branches to HEADs of corresponding temporary branches.
+    // The last step may fail if there are concurrent commits to source branch, in which case
+    // the whole process is re-tried.
+    try (NessieApiV2 api = createNessieApiInstance()) {
+      rootHash = api.getConfig().getNoAncestorHash();
+      for (int attempt = 0; attempt < maxAttempts; attempt++) {
+        spec.commandLine().getOut().printf("Running attempt %d...%n", attempt);
+
+        sources.clear();
+        targets.clear();
+        tmpSuffix = "-tmp-" + UUID.randomUUID();
+
+        super.execute(); // copy content into temporary branches
+
+        try {
+          // reassign original branch names to HEADs of corresponding temporary branches
+          reassign(api);
+
+          spec.commandLine().getOut().printf("Completed successfully%n");
+          break;
+        } catch (NessieReferenceConflictException e) {
+          spec.commandLine()
```

Comment:
In case of retries, existing target branches, created by the previous retry attempt, will be considered source branches for the next retry attempt. Then the logic goes on to migrate the target branch to a newly-created target branch whose name is something like `main-tmp-842f08fc-3d89-4665-bc86-a3f0bab0a09e-tmp-e3193ff1-8c28-483d-aacd-506d76b4baac`.

I think you want to delete all target branches here.

```suggestion
          for (Reference target : targets.values()) {
            target = api.getReference().refName(target.getName()).get();
            api.deleteReference().reference(target).delete();
          }
          spec.commandLine()
```

Reply:
nice catch, @adutra ! fixed.